src/python/pants/backend/python/util_rules/scm_versioning.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from typing import cast
+
+import toml
+
+from pants.backend.python.subsystems.setuptools_scm import SetuptoolsSCM
+from pants.backend.python.target_types import (
+    PythonSourceField,
+    SetuptoolsSCMDummySourceField,
+    SetuptoolsSCMTagRegexField,
+    SetuptoolsSCMWriteToField,
+    SetuptoolsSCMWriteToTemplateField,
+)
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.engine.fs import CreateDigest, FileContent
+from pants.engine.internals.native_engine import Digest, Snapshot
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import GeneratedSources, GenerateSourcesRequest
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.vcs.git import GitWorktreeRequest, MaybeGitWorktree
+
+
+class SCMVersioningError(Exception):
+    pass
+
+
+# Note that even though setuptools_scm is Python-centric, we could easily use it to generate
+# version data for use in other languages!
+class GeneratePythonFromSetuptoolsSCMRequest(GenerateSourcesRequest):
+    input = SetuptoolsSCMDummySourceField
+    output = PythonSourceField
+
+
+@rule
+async def generate_python_from_setuptools_scm(
+    request: GeneratePythonFromSetuptoolsSCMRequest,
+    setuptools_scm: SetuptoolsSCM,
+) -> GeneratedSources:
+    # A GitWorktreeRequest is uncacheable, so this enclosing rule will run every time its result
+    # is needed, meaning it will always return a result based on the current underlying git state.
+    maybe_git_worktree = await Get(MaybeGitWorktree, GitWorktreeRequest())
+    if not maybe_git_worktree.git_worktree:
+        raise SCMVersioningError("Not running in a git worktree")
```

Comment:
Should probably mention: the `setuptools_scm_version` target type in here.

Reply:
Agreed. Something like

> Trying to determine the version for the `scm_version` target `{address}`, but you are not running in a Git worktree.

Then maybe a remedy like:

> Perhaps run `git init` or make sure you ran `git clone` correctly?

Reply:
Done, but I'm not sure about the remedy, since there could be all sorts of reasons, from "this repo doesn't use Git" to "this is running on an unzipped code archive". So I think I will leave that to the user.